packages/service/src/utils/documents.ts

```diff
@@ -19,15 +19,33 @@ export const createDocument = ({
   doc: Y.Doc;
   user: { id: string };
 }) =>
-  db
-    .insert("document", {
-      value: doc.getText(TEXT_NAME).toJSON(),
-      origin: Buffer.from(Y.encodeStateAsUpdate(doc)),
-      web_rtc_key: "",
-      creator_id: user.id,
-      latest_update_time: new Date(),
-    })
-    .run(pool);
+  db.serializable(pool, async (txnClient) => {
+    try {
+      await db.sql`SAVEPOINT start`.run(txnClient);
+      const dbDoc = await db
+        .insert("document", {
+          value: doc.getText(TEXT_NAME).toJSON(),
+          origin: Buffer.from(Y.encodeStateAsUpdate(doc)),
+          web_rtc_key: "",
+          creator_id: user.id,
+          latest_update_time: new Date(),
+        })
+        .run(txnClient);
+      return {
+        document: dbDoc,
+        user_document: await db
+          .insert("user_document", {
+            document_id: dbDoc.id,
+            user_id: user.id,
+          })
+          .run(txnClient),
+      };
+    } catch (err) {
+      if (!db.isDatabaseError(err)) throw err;
+      await db.sql`ROLLBACK TO start`.run(txnClient);
```

Comment:
I think zapatos handles this for us, we shouldn't need to write this again

see https://github.com/jawj/zapatos/blob/master/src/db/transaction.ts#L99